components/butterfly/src/rumor.rs

```diff
@@ -408,17 +477,60 @@ impl<T> RumorStore<T> where T: Rumor
     /// We don't care if this repeats - it just needs to be unique for any given two states, which
     /// it will be.
     fn increment_update_counter(&self) { self.update_counter.fetch_add(1, Ordering::Relaxed); }
+
+    pub fn keys(&self) -> Vec<RumorKey> {
+        self.read_entries()
+            .values()
+            .flat_map(HashMap::values)
+            .map(RumorKey::from)
+            .collect()
+    }
+}
+
+impl<T> RumorStore<T> where T: Transient
+{
+    /// Find rumors in our rumor store that have expired.
+    fn partitioned<B>(&self, expiration_date: DateTime<Utc>) -> (B, B)
+        where B: Default + Extend<T>
+    {
+        self.read_entries()
+            .values()
+            .flat_map(HashMap::values)
+            .cloned()
+            .partition(|rumor| rumor.expiration().expired(expiration_date))
+    }
+
+    pub fn expired(&self, expiration_date: DateTime<Utc>) -> Vec<T> {
+        self.partitioned(expiration_date).0
+    }
+
+    pub fn live(&self, expiration_date: DateTime<Utc>) -> Vec<T> {
+        self.partitioned(expiration_date).1
+    }
+
+    /// Remove all rumors that have expired from our rumor store.
+    pub fn purge_expired(&self, expiration_date: DateTime<Utc>) {
+        self.expired(expiration_date).iter().for_each(|r| {
+                                                self.remove(r.key(), r.id());
+                                            });
+    }
```

Comment:
It doesn't seem that anything that ultimately calls this passes anything other than "now" as an expiration date. By pulling that detail internal to the implementation, we simplify life for the callers.

If you want to retain the flexibility for testing purposes, that could be factored into the internal implementation. It doesn't seem that the public interface benefits much from passing in an `expiration_date`, though.

Reply:
One downside to making that detail internal to this function is that every caller of this function gets a slightly different version of `now`.

In `expire.rs`, for example, we create a single `DateTime` instance that represents `now` and pass that into this function for all the various rumor stores.  That ensures that every invocation of this for the various rumor stores is dealing with the same expiration date.  Removing that parameter would mean that every rumor store would be operating with slightly different expiration dates.

In practice, it's probably fine most of the time, but that feels like the kind of detail that could easily result in hard to debug corner cases.  All things considered, I think I'd rather leave the expiration date as a parameter.  I don't see any downside to doing so.